cli/command/service/trust.go

```diff
@@ -0,0 +1,100 @@
+package service
+
+import (
+	"encoding/hex"
+	"fmt"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/distribution/digest"
+	distreference "github.com/docker/distribution/reference"
+	"github.com/docker/docker/api/types/swarm"
+	"github.com/docker/docker/cli/command"
+	"github.com/docker/docker/cli/trust"
+	"github.com/docker/docker/reference"
+	"github.com/docker/docker/registry"
+	"github.com/docker/notary/tuf/data"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+)
+
+func resolveServiceImageDigest(dockerCli *command.DockerCli, service *swarm.ServiceSpec) {
+	if !command.IsTrusted() {
+		// Digests are resolved by the daemon when not using content
+		// trust.
+		return
+	}
+
+	image := service.TaskTemplate.ContainerSpec.Image
+
+	// We only attempt to resolve the digest if the reference
+	// could be parsed as a digest reference. Specifying an image ID
+	// is valid but not resolvable. There is no warning message for
+	// an image ID because it's valid to use one.
+	if _, err := digest.ParseDigest(image); err == nil {
+		return
+	}
+
+	ref, err := reference.ParseNamed(image)
+	if err != nil {
+		fmt.Fprintf(dockerCli.Err(), "Warning: could not parse image reference %s\n", service.TaskTemplate.ContainerSpec.Image)
+		return
+	}
+	if _, ok := ref.(reference.Canonical); !ok {
+		if reference.IsNameOnly(ref) {
```

Comment:
Non-blocking:  I didn't know what this did, so I looked it up, and it looks like the current vendored version of distribution's `WithDefaultTag` checks `IsNameOnly` before adding the default tag, so possibly this extra `reference.IsNameOnly(ref)` check is not necessary?